app/src/main/java/io/github/hidroh/materialistic/data/FileDownloader.java

```diff
@@ -0,0 +1,57 @@
+package io.github.hidroh.materialistic.data;
+
+import android.content.Context;
+import android.os.Handler;
+import android.os.Looper;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.Request;
+import okhttp3.Response;
+import okio.BufferedSink;
+import okio.Okio;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+
+public class FileDownloader {
+    private Call.Factory mCallFactory;
+    private final String mCacheDir;
+
+    @Inject
+    public FileDownloader(Context context, Call.Factory callFactory) {
+        mCacheDir = context.getCacheDir().getPath(); // don't need to keep a reference to context after this
+        mCallFactory = callFactory;
+    }
+
+    public void downloadFile(String url, String mimeType, FileDownloaderCallback callback) {
+        final Request request = new Request.Builder().url(url)
+                .addHeader("Content-Type", mimeType)
+                .build();
+
+        mCallFactory.newCall(request).enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException e) {
+                callback.onFailure(e);
```

Comment:
This needs to be wrapped with `new Handler(Looper.getMainLooper()).post(() -> callback.onFailure(e))` as well.